### PR TITLE
New example "Reading specified resources from YAML file"

### PR DESCRIPTION
**Pull request by stefanhans**
**What this PR does / why we need it**:

The PR shows how to read a simple resource specification in YAML and to process it until the creation and/or update has been done.

**Which issue(s) this PR fixes**

The PR does not fix any issue, but it shows an example.

**Special notes for the reviewer**:

Please check, if the proceeding is as it should be.

**Release note**:

New client-go example showing how to read simple deployment with service YAML configuration and process accordingly.

### CODE REVIEW

**Comment by stefanhans** [comment]
Hi,

I've made a PR to add an example for the client-go. Currently, I try to understand what I have to do to get the pull-kubernetes-verify job done successfully.

This is my first PR and I'm a bit confused about all. Maybe somebody could give me a hint what to do before I have a try and do possibly something useless or stupid.

Cheers, Stefan

**Comment by stefanhans** [comment]
Hi Benjamin,

thank you for your answer. BTW: Should I reply to all?

Aside from not really understanding what it is all about, 🤔 I've got the
following output:

╭─[~/GolandProjects/kubernetes]─[stefan@Stefans-MBP]─[1]─[504]

╰─[:(] % ./hack/update-bazel.sh

../../internal/config/directives.go:24:2: cannot find package "
github.com/bazelbuild/buildtools/build" in any of:

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.VDrBQKdE/go/src/
github.com/bazelbuild/bazel-gazelle/vendor/github.com/bazelbuild/buildtools/build
(vendor tree)

/usr/local/go/src/github.com/bazelbuild/buildtools/build (from $GOROOT)

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.VDrBQKdE/go/src/
github.com/bazelbuild/buildtools/build (from $GOPATH)

../../internal/rules/construct.go:27:2: cannot find package "
github.com/bazelbuild/buildtools/tables" in any of:

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.VDrBQKdE/go/src/
github.com/bazelbuild/bazel-gazelle/vendor/github.com/bazelbuild/buildtools/tables
(vendor tree)

/usr/local/go/src/github.com/bazelbuild/buildtools/tables (from $GOROOT)

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.VDrBQKdE/go/src/
github.com/bazelbuild/buildtools/tables (from $GOPATH)

!!! [0314 09:47:49] Call tree:

!!! [0314 09:47:49]  1: ./hack/update-bazel.sh:34
kube::util::go_install_from_commit(...)

!!! Error in /Users/stefan/GolandProjects/kubernetes/hack/lib/util.sh:473

  Error in /Users/stefan/GolandProjects/kubernetes/hack/lib/util.sh:473.
'((i<3-1))' exited with status 1

Call stack:

  1: /Users/stefan/GolandProjects/kubernetes/hack/lib/util.sh:473
kube::util::go_install_from_commit(...)

  2: ./hack/update-bazel.sh:34 main(...)

Exiting with status 1

╭─[~/GolandProjects/kubernetes]─[stefan@Stefans-MBP]─[1]─[505]

╰─[:(] % bazel version

Build label: 0.11.1-homebrew

Build target:
bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar

Build time: Sun May 31 16:57:39 +50150 (1520426998659)

Build timestamp: 1520426998659

Build timestamp as int: 1520426998659

╭─[~/GolandProjects/kubernetes]─[stefan@Stefans-MBP]─[0]─[506]

╰─[:)] %


After


git clone https://github.com/bazelbuild/buildtools.git

cd buildtools/build

go install




╭─[~/go/src/k8s.io/kubernetes]─[stefan@Stefans-MBP]─[0]─[597]

╰─[:)] % ./hack/update-bazel.sh

╭─[~/go/src/k8s.io/kubernetes]─[stefan@Stefans-MBP]─[0]─[597]




oops, I did it in the wrong directory 🙁


In the right directory it does not work - with same GOROOT and GOPATH - I
decided to "hack it"


sudo cp -r /Users/stefan/go/src/github.com/bazelbuild /usr/local/go/src/
github.com

and the error changed


╭─[~/GolandProjects/kubernetes]─[stefan@Stefans-MBP]─[0]─[639]

╰─[:)] % ./hack/update-bazel.sh


F0314 12:11:11.704128   32126 kazel.go:76] err walking generated:
openapi-gen path outside of k8s.io/kubernetes: .

goroutine 1 [running]:

github.com/kubernetes/repo-infra/vendor/github.com/golang/glog.stacks(0xc4204e7c00,
0xc42001c0d0, 0x73, 0xc7)

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.LdVjqFyc/go/src/
github.com/kubernetes/repo-infra/vendor/github.com/golang/glog/glog.go:769
+0xcf

github.com/kubernetes/repo-infra/vendor/github.com/golang/glog.(*loggingT).output(0x12d24a0,
0xc400000003, 0xc4203a60b0, 0x12b48a2, 0x8, 0x4c, 0x0)

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.LdVjqFyc/go/src/
github.com/kubernetes/repo-infra/vendor/github.com/golang/glog/glog.go:720
+0x32d

github.com/kubernetes/repo-infra/vendor/github.com/golang/glog.(*loggingT).printf(0x12d24a0,
0x3, 0x11d79de, 0x19, 0xc42017bee8, 0x1, 0x1)

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.LdVjqFyc/go/src/
github.com/kubernetes/repo-infra/vendor/github.com/golang/glog/glog.go:655
+0x14b

github.com/kubernetes/repo-infra/vendor/github.com/golang/glog.Fatalf(0x11d79de,
0x19, 0xc42017bee8, 0x1, 0x1)

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.LdVjqFyc/go/src/
github.com/kubernetes/repo-infra/vendor/github.com/golang/glog/glog.go:1148
+0x67

main.main()

/var/folders/x4/gf3sv2rs6vb46l0k6ht09vn00000gn/T/tmp.LdVjqFyc/go/src/
github.com/kubernetes/repo-infra/kazel/kazel.go:76 +0x29d

!!! Error in ./hack/update-bazel.sh:51

  Error in ./hack/update-bazel.sh:51. 'kazel' exited with status 255

Call stack:

  1: ./hack/update-bazel.sh:51 main(...)

Exiting with status 1

╭─[~/GolandProjects/kubernetes]─[stefan@Stefans-MBP]─[1]─[640]

╰─[:(] %

and the BUILD file changed as well which I was pushing desperately


Greetings from a feeling stupid and confused chaos monkey 😵

**Comment by BenTheElder** [comment]
Scanning your post it looks like your kubernetes checkout is not in `$GOPATH/src/k8s.io/kubernetes`. A number of things expect this path.
Ref: https://github.com/thtanaka/kubernetes/blob/master/docs/devel/development.md#clone-your-fork

If this is the case, then setting this up correctly should let the scripts work. You should not need to install bazel or the build tools yourself to get things working after this.

As an aside if you place log output between three ` before and after [GitHub will nicely format them](https://guides.github.com/features/mastering-markdown/) which is helpful.

Hope that helps! If not please ping me with an update and I'll try to help further.

**Comment by BenTheElder** [comment]
Er following up it looks like you eventually got things sorted 🙃
You need to put the release not info in your PR body to 
\`\`\`release-note
NONE
\`\`\`
And then we just need one of rhe owners to review and approve. (See the bot message above)

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		panic(err)
+	}
+
+	// Split YAML into chunks or k8s resources, respectively
+	yamlDecoder := yaml.NewDocumentDecoder(ioutil.NopCloser(reader))
+
+	// Create decoding function used for YAML to JSON decoding
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+
+	// Read first resource - expecting deployment with size < 2048
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: and if it is bigger?

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	yamlDeployment := make([]byte, 2048)
+	_, err = yamlDecoder.Read(yamlDeployment)
+	if err != nil {
+		panic(err)
+	}
+
+	// Trim unnecessary trailing 0x0 signs which are not accepted
+	TrimmedYamlDeployment := strings.TrimRight(string(yamlDeployment), string(byte(0)))
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: local vars are usually lower case

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+	// Decode deployment resource from YAML to JSON
+	jsonDeployment, _, err := decode([]byte(TrimmedYamlDeployment), nil, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	// Check "kind: deployment"
+	if jsonDeployment.GetObjectKind().GroupVersionKind().Kind != "Deployment" {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: why?

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+	// Marshall JSON deployment
+	d, err := json.Marshal(&jsonDeployment)
+	if err != nil {
+		panic(err)
+	}
+
+	// Unmarshall JSON into deployment struct
+	var deployment appsv1beta1.Deployment
+	err = json.Unmarshal(d, &deployment)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: wasn't jsonDeployment already of this type?

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	// Create client for deployments
+	deploymentsClient := clientset.AppsV1beta1().Deployments(corev1.NamespaceDefault)
+
+	// Create client for services
+	servicesClient := clientset.CoreV1().Services(corev1.NamespaceDefault)
+
+	// create deployment - try update on error
+	fmt.Printf("Create deployment %q\n", deployment.Name)
+	createdDeployment, err := deploymentsClient.Create(&deployment)
+	if err != nil {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: this should check the error. Only update if it existed already.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+	runningService, err := servicesClient.Get(service.Name, metav1.GetOptions{})
+	if err != nil {
+		panic(err)
+	}
+
+	// Get Pod "kube-addon-manager-minikube" of "kube-system" to retrieve 'minikube ip'
+	//
+	// Only valid for a minikube environment
+	//
+	pod, err := clientset.CoreV1().Pods("kube-system").Get("kube-addon-manager-minikube", metav1.GetOptions{})
```
Comment (sttts) [inline staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: this is not portable. The example should work in every environment.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 1 above)
Comment (stefanhans) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: I take the size of the file. Not perfect, but I don't know a better way so far.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 3 above)
Comment (stefanhans) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: I've removed it. Better would be if we could act related to the groupVersionKind in a more generalized way. Can you give me a hint?

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 4 above)
Comment (stefanhans) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: I couldn't find out how to create the 'container' for "groupVersionKind' without knowing beforehand.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 5 above)
Comment (stefanhans) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: I think I cannot reliably match the error string to figure out the 'is already running' error. Instead, I would try to check it out in advance.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 6 above)
Comment (stefanhans) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: No idea how to find out the IP address generally. Let's see what I can investigate tomorrow.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: there are helpers like `k8s.io/apimachinery/pkg/api/errors.IsNotFound`

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: should be dynamic, i.e. no fixed size at all.

File: staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/examples/read-yaml-configuration-deployment/main.go]: The object created by the codec should be the Deployment. You can type cast it.

**Pull request closed without merging**